alphagov/paas-cf
# [#114308641] Insert pre-generated/purchased public certificates

[#114308641 Insert public-facing certs into Pipeline](https://www.pivotaltracker.com/story/show/114308641)

## What

We need to have some way to inject the certs and corresponding private keys that tenants and the public will use into an environment.

These certs, for the apps wildcard and the CF endpoints that tenants use, will be signed by a CA, so the process looks a bit different in comparison to the ones generated inside the pipeline. Outside the pipeline we'll be creating a key, generating a CSR and getting the signed cert outside our pipeline.

The operator deploying the environment might not have access and permissions to read the keys, so we decided to split the process to upload the certificates to AWS.

## Dependencies

After merging this PR, the environments with specific certificates require some manual intervention to do the initial manual upload of the certificates.

Note that after merging the PR, we must run the manual step `make manually_upload_certs` o upload the certificates in CI, stage and prod. 

Meanwhile we don't have the final certificates for CI, stage and prod we will use the self-signed generated ones that are in the repository. They are stored in [a branch of the `paas-pass`  password store](https://github.gds/government-paas/credentials/pull/13), so the tasks to run will be:

```
paas-pass git fetch
paas-pass git checkout insert-prod-certs-114308641

export CERT_PASSWORD_STORE_DIR=~/.paas-pass

# Set AWS creds for CI
make ci manually_upload_certs DEPLOY_ENV=master

# Get AWS creds for Stage
make stage manually_upload_certs DEPLOY_ENV=staging

# Get AWS creds for Prod
make prod manually_upload_certs DEPLOY_ENV=prod
```

## How to review

The PR must be reviewed in this context:
 * I want to keep uploading self-signed generated certificates in development environments.
 * I want to be able to manually upload pregenerated certificates for non development environments (prod/stage)
 * I want to consume the purchased certificate from a password store like `paas-pass` (**Note: The certificate store might change in the future**)

## How to test this:

> Note: For testing it is not necessary to deploy all cloudfoundry, it is enough only run up to `cf-terraform` and check that the certificates have been generated properly.

To test:

 1. Deploy your environment as usual, it should work (you can consider OK if it gets up to `cf-deploy` task)
 2. Download the router certificates and store them in a `paas-pass` for testing:

     ```
DEPLOY_ENV=...
AWS_ACCOUNT=dev

aws s3 cp s3://${DEPLOY_ENV}-state/cf-certs.tar.gz /tmp --region eu-west-1
mkdir /tmp/cf-certs
tar -xvzf /tmp/cf-certs.tar.gz -C /tmp/cf-certs
paas-pass insert -m certs/${AWS_ACCOUNT}/${DEPLOY_ENV}/router_external.key < /tmp/cf-certs/router_external.key
paas-pass insert -m certs/${AWS_ACCOUNT}/${DEPLOY_ENV}/router_external.crt < /tmp/cf-certs/router_external.crt
```
 3. Try to upload these certificates manually:

     ```
make dev manually_upload_certs CERT_PASSWORD_STORE_DIR=~/.paas-pass
```
 4. Run the deployment but setting `SKIP_UPLOAD_GENERATED_CERTS=true`:

     ```
SKIP_UPLOAD_GENERATED_CERTS=true BRANCH=$(git rev-parse --abbrev-ref HEAD) make ${AWS_ACCOUNT} pipelines DEPLOY_ENV=...
```
 5. Finally: Delete the deployment using `destroy-cloudfoundry`. It should delete the certificates.

 6. Follow the instructions in  `doc/non_dev_deployments.md` to upload the certs manually and rerun the deployment (with `SKIP_UPLOAD_GENERATED_CERTS=true` as mention before).

## Who can review

Anyone but @keymon or @mtekel
